base/dates/parse.jl

```diff
@@ -1,97 +1,132 @@
-    tuple_type = slot_types(T)
 
-    # `slot_order`, `slot_defaults`, and `slot_types` return tuples of the same length
-    assert(num_types == length(field_order) == length(field_defaults))
+@generated function tryparse_core(str::AbstractString, df::DateFormat, raise::Bool=false)
+    token_types, directive_index, directive_letters = directives(df)
+
+    directive_types = Type[FORMAT_SPECIFIERS[letter] for letter in directive_letters]
+    directive_names = Symbol[genvar(t) for t in directive_types]
+    directive_defaults = Tuple(FORMAT_DEFAULTS[t] for t in directive_types)
+    R = typeof(directive_defaults)
+
+    # Pre-assign output variables to default values. Allows us to use `@goto done` without
+    # worrying about unassigned variables.
+    assign_defaults = Expr[
+        quote
+            $name = $default
+        end
+        for (name, default) in zip(directive_names, directive_defaults)
+    ]
+
+    parsers = Expr[
+        begin
+            di = directive_index[i]
+            if di != 0
+                name = directive_names[di]
+                nullable = Symbol(:nullable_, name)
+                quote
+                    pos > len && @goto done
+                    $nullable, next_pos = tryparsenext(tokens[$i], str, pos, len, locale)
+                    isnull($nullable) && @goto error
+                    $name = unsafe_get($nullable)
+                    pos = next_pos
+                    directive_idx += 1
+                    token_idx += 1
+                end
+            else
+                quote
+                    pos > len && @goto done
+                    nullable_delim, next_pos = tryparsenext(tokens[$i], str, pos, len, locale)
+                    isnull(nullable_delim) && @goto error
+                    pos = next_pos
+                    token_idx += 1
+                end
+            end
+        end
+        for i in 1:length(token_types)
+    ]
 
     quote
-        R = Nullable{$tuple_type}
-        t = df.tokens
-        l = df.locale
+        tokens = df.tokens
+        locale::DateLocale = df.locale
         pos, len = start(str), endof(str)
+        directive_idx = 0
+        token_idx = 1
+
+        $(assign_defaults...)
+        $(parsers...)
 
-        err_idx = 1
-        Base.@nexprs $N i->val_i = 0
-        Base.@nexprs $N i->(begin
-            pos > len && @goto done
-            nv, next_pos = tryparsenext(t[i], str, pos, len, l)
-            isnull(nv) && @goto error
-            val_i, pos = unsafe_get(nv), next_pos
-            err_idx += 1
-        end)
-        pos <= len && @goto error
+        pos > len || @goto error
 
         @label done
-        parts = Base.@ntuple $N val
-        return R(reorder_args(parts, $field_order, $field_defaults, err_idx)::$tuple_type)
+        return Nullable{$R}($(Expr(:tuple, directive_names...))), directive_idx
 
         @label error
         # Note: Keeping exception generation in separate function helps with performance
```

Comment:
I tested this again and it no longer seems to be an issue. I originally found this problem.